File: app/src/main/java/com/pbrlm/abhaya/domain/service/PriorityCalculator.kt
package com.pbrlm.abhaya.domain.service

import com.pbrlm.abhaya.domain.model.EmergencyPriority
import com.pbrlm.abhaya.domain.model.EmergencyType
import com.pbrlm.abhaya.domain.model.MedicalEmergencyType
import com.pbrlm.abhaya.domain.model.WaterDepth

/**
 * Isolated, testable priority calculation service.
 * All priority logic lives here — never scattered across the UI or ViewModels.
 *
 * Rules:
 * CRITICAL: Immediate life threat — trapped, above-chest water, rising water, severe medical,
 *           unconscious, severe bleeding, multiple people in danger, vulnerable people in danger.
 * HIGH: Serious rescue needed — boat rescue, multiple people, water at chest/waist level.
 * MEDIUM: Shelter, non-urgent assistance.
 * LOW: Other non-critical requests.
 */
object PriorityCalculator {

    fun calculateForSos(): EmergencyPriority = EmergencyPriority.CRITICAL

    fun calculateForWaterRescue(
        waterDepth: WaterDepth,
        waterRising: Boolean?,
        peopleTrapped: Boolean,
        injured: Boolean,
        numberOfPeople: Int,
        children: Boolean,
        elderly: Boolean,
        disabled: Boolean
    ): EmergencyPriority {
        // Immediate CRITICAL conditions
        if (peopleTrapped) return EmergencyPriority.CRITICAL
        if (waterDepth == WaterDepth.ABOVE_CHEST) return EmergencyPriority.CRITICAL
        if (waterRising == true && waterDepth == WaterDepth.CHEST) return EmergencyPriority.CRITICAL
        if (waterRising == true && waterDepth == WaterDepth.WAIST && (children || elderly || disabled)) {
            return EmergencyPriority.CRITICAL
        }
        if (injured) return EmergencyPriority.CRITICAL
        if (numberOfPeople >= 5 && waterDepth == WaterDepth.CHEST) return EmergencyPriority.CRITICAL

        // HIGH conditions
        if (waterDepth == WaterDepth.CHEST) return EmergencyPriority.HIGH
        if (waterRising == true) return EmergencyPriority.HIGH
        if (numberOfPeople >= 3) return EmergencyPriority.HIGH
        if (children || elderly || disabled) return EmergencyPriority.HIGH
        if (waterDepth == WaterDepth.WAIST) return EmergencyPriority.HIGH

        // MEDIUM
        if (waterDepth == WaterDepth.KNEE || waterDepth == WaterDepth.ANKLE) {
            return EmergencyPriority.MEDIUM
        }

        return EmergencyPriority.HIGH // Default for water — always serious
    }

    fun calculateForMedical(
        medicalType: MedicalEmergencyType,
        personUnconscious: Boolean,
        breathingProblem: Boolean,
        severeBleeding: Boolean,
        pregnancyRelated: Boolean,
        injured: Boolean,
        numberOfPeople: Int,
        children: Boolean,
        elderly: Boolean,
        disabled: Boolean
    ): EmergencyPriority {
        // CRITICAL conditions
        if (personUnconscious) return EmergencyPriority.CRITICAL
        if (breathingProblem) return EmergencyPriority.CRITICAL
        if (severeBleeding) return EmergencyPriority.CRITICAL
        if (medicalType == MedicalEmergencyType.UNCONSCIOUS) return EmergencyPriority.CRITICAL
        if (medicalType == MedicalEmergencyType.BREATHING_PROBLEM) return EmergencyPriority.CRITICAL
        if (medicalType == MedicalEmergencyType.SEVERE_BLEEDING) return EmergencyPriority.CRITICAL
        if (numberOfPeople >= 3 && injured) return EmergencyPriority.CRITICAL

        // HIGH conditions
        if (pregnancyRelated || medicalType == MedicalEmergencyType.PREGNANCY) {
            return EmergencyPriority.HIGH
        }
        if (injured) return EmergencyPriority.HIGH
        if (children || elderly || disabled) return EmergencyPriority.HIGH
        if (numberOfPeople >= 3) return EmergencyPriority.HIGH

        // MEDIUM
        return EmergencyPriority.MEDIUM
    }

    fun calculateForBoatRescue(
        waterDepth: WaterDepth,
        waterRising: Boolean?,
        peopleTrapped: Boolean,
        hasmedicalEmergency: Boolean,
        numberOfPeople: Int,
        children: Boolean,
        elderly: Boolean,
        disabled: Boolean
    ): EmergencyPriority {
        if (peopleTrapped) return EmergencyPriority.CRITICAL
        if (hasmedicalEmergency) return EmergencyPriority.CRITICAL
        if (waterDepth == WaterDepth.ABOVE_CHEST) return EmergencyPriority.CRITICAL
        if (waterRising == true && (children || elderly || disabled)) return EmergencyPriority.CRITICAL

        if (waterDepth == WaterDepth.CHEST) return EmergencyPriority.HIGH
        if (waterRising == true) return EmergencyPriority.HIGH
        if (numberOfPeople >= 5) return EmergencyPriority.HIGH
        if (children || elderly || disabled) return EmergencyPriority.HIGH

        return EmergencyPriority.HIGH // Boat rescue is always at least HIGH
    }

    fun calculateForShelter(
        numberOfPeople: Int,
        children: Boolean,
        elderly: Boolean,
        disabled: Boolean
    ): EmergencyPriority {
        if (children || elderly || disabled) return EmergencyPriority.HIGH
        if (numberOfPeople >= 5) return EmergencyPriority.MEDIUM
        return EmergencyPriority.MEDIUM
    }

    /**
     * Determines if escalation is required.
     * Criteria: CRITICAL priority + ACTIVE status + unresolved for > 30 minutes.
     */
    fun isEscalationRequired(
        priority: EmergencyPriority,
        status: com.pbrlm.abhaya.domain.model.EmergencyStatus,
        createdAtMillis: Long,
        currentTimeMillis: Long = System.currentTimeMillis()
    ): Boolean {
        if (priority != EmergencyPriority.CRITICAL) return false
        if (status != com.pbrlm.abhaya.domain.model.EmergencyStatus.ACTIVE &&
            status != com.pbrlm.abhaya.domain.model.EmergencyStatus.ACKNOWLEDGED
        ) return false
        val elapsedMinutes = (currentTimeMillis - createdAtMillis) / 60_000
        return elapsedMinutes >= 30
    }
}

File: AbhayaRN/src/services/priorityCalculator.ts
import type {
  EmergencyPriority,
  EmergencyStatus,
  MedicalEmergencyType,
  WaterDepth,
} from "../types/emergency";

export function calculateSosPriority(): EmergencyPriority {
  return "CRITICAL";
}

export function calculateWaterRescuePriority(params: {
  waterDepth: WaterDepth;
  waterRising?: boolean;
  peopleTrapped: boolean;
  injured: boolean;
  numberOfPeople: number;
  children: boolean;
  elderly: boolean;
  disabled: boolean;
}): EmergencyPriority {
  const { waterDepth, waterRising, peopleTrapped, injured, numberOfPeople, children, elderly, disabled } = params;

  if (peopleTrapped) return "CRITICAL";
  if (waterDepth === "ABOVE_CHEST") return "CRITICAL";
  if (waterRising && waterDepth === "CHEST") return "CRITICAL";
  if (waterRising && waterDepth === "WAIST" && (children || elderly || disabled)) return "CRITICAL";
  if (injured) return "CRITICAL";
  if (numberOfPeople >= 5 && waterDepth === "CHEST") return "CRITICAL";

  if (waterDepth === "CHEST") return "HIGH";
  if (waterRising) return "HIGH";
  if (numberOfPeople >= 3) return "HIGH";
  if (children || elderly || disabled) return "HIGH";
  if (waterDepth === "WAIST") return "HIGH";
  if (waterDepth === "KNEE" || waterDepth === "ANKLE") return "MEDIUM";

  return "HIGH";
}

export function calculateMedicalPriority(params: {
  medicalType: MedicalEmergencyType;
  personUnconscious: boolean;
  breathingProblem: boolean;
  severeBleeding: boolean;
  pregnancyRelated: boolean;
  injured: boolean;
  numberOfPeople: number;
  children: boolean;
  elderly: boolean;
  disabled: boolean;
}): EmergencyPriority {
  const { personUnconscious, breathingProblem, severeBleeding, medicalType, injured, numberOfPeople, children, elderly, disabled, pregnancyRelated } = params;

  if (personUnconscious) return "CRITICAL";
  if (breathingProblem) return "CRITICAL";
  if (severeBleeding) return "CRITICAL";
  if (medicalType === "UNCONSCIOUS" || medicalType === "BREATHING_PROBLEM" || medicalType === "SEVERE_BLEEDING") return "CRITICAL";
  if (numberOfPeople >= 3 && injured) return "CRITICAL";

  if (pregnancyRelated || medicalType === "PREGNANCY") return "HIGH";
  if (injured) return "HIGH";
  if (children || elderly || disabled) return "HIGH";
  if (numberOfPeople >= 3) return "HIGH";

  return "MEDIUM";
}

export function calculateBoatRescuePriority(params: {
  waterDepth: WaterDepth;
  waterRising?: boolean;
  peopleTrapped: boolean;
  medicalEmergency: boolean;
  numberOfPeople: number;
  children: boolean;
  elderly: boolean;
  disabled: boolean;
}): EmergencyPriority {
  const { peopleTrapped, medicalEmergency, waterDepth, waterRising, children, elderly, disabled } = params;

  if (peopleTrapped) return "CRITICAL";
  if (medicalEmergency) return "CRITICAL";
  if (waterDepth === "ABOVE_CHEST") return "CRITICAL";
  if (waterRising && (children || elderly || disabled)) return "CRITICAL";

  if (waterDepth === "CHEST") return "HIGH";
  if (waterRising) return "HIGH";
  if (children || elderly || disabled) return "HIGH";

  return "HIGH";
}

export function calculateShelterPriority(params: {
  numberOfPeople: number;
  children: boolean;
  elderly: boolean;
  disabled: boolean;
}): EmergencyPriority {
  const { children, elderly, disabled } = params;
  if (children || elderly || disabled) return "HIGH";
  return "MEDIUM";
}

export function isEscalationRequired(
  priority: EmergencyPriority,
  status: EmergencyStatus,
  createdAtMs: number,
  nowMs: number = Date.now()
): boolean {
  if (priority !== "CRITICAL") return false;
  if (status !== "ACTIVE" && status !== "ACKNOWLEDGED") return false;
  const elapsedMinutes = (nowMs - createdAtMs) / 60_000;
  return elapsedMinutes >= 30;
}
